stdlib/asm/crypto/hashes/sha256.masm

```diff
@@ -0,0 +1,3096 @@
+    # stack = [a + 0, a + 1, a + 2, a + 3, ...] #
+end
+
+proc.reorder_stack_words
+    swapw
+    movupw.3
+    movupw.2
+    movupw.3
+end
+
+# SHA256 function; see https://github.com/itzmeanjan/merklize-sha/blob/8a2c006a2ffe1e6e8e36b375bc5a570385e9f0f2/include/sha2.hpp#L89-L113 #
+proc.prepare_message_schedule.5
+    popw.local.0
+    popw.local.1
+    popw.local.2
+    popw.local.3
+
+    movupw.3
+    movupw.3
+
+    # --- #
+
+    exec.gen_four_message_words
+
+    popw.local.4
+    movupw.2
+
+    pushw.local.0
+    repeat.3
+        swap
+        drop
+    end
+
+    popw.mem # write to mem msg[0, 1, 2, 3] #
+    pushw.local.4
+
+    exec.reorder_stack_words
+    
+    # --- #
+
+    exec.gen_four_message_words
+
+    popw.local.4
+    movupw.2
+
+    pushw.local.0
+    drop
+    repeat.2
+        swap
+        drop
+    end
+
+    popw.mem # write to mem msg[4, 5, 6, 7] #
+    pushw.local.4
+
+    exec.reorder_stack_words
+
+    # --- #
+
+    exec.gen_four_message_words
+
+    popw.local.4
+    movupw.2
+
+    pushw.local.0
+    drop
+    drop
+    swap
+    drop
+
+    popw.mem # write to mem msg[8, 9, 10, 11] #
+    pushw.local.4
+
+    exec.reorder_stack_words
+
+    # --- #
+
+    exec.gen_four_message_words
+
+    popw.local.4
+    movupw.2
+
+    pushw.local.0
+    drop
+    drop
+    drop
+
+    popw.mem # write to mem msg[12, 13, 14, 15] #
+    pushw.local.4
+
+    exec.reorder_stack_words
+
+    # --- #
+    # --- #
+
+    exec.gen_four_message_words
+
+    popw.local.4
+    movupw.2
+
+    pushw.local.1
+    repeat.3
+        swap
+        drop
+    end
+
+    popw.mem # write to mem msg[16, 17, 18, 19] #
+    pushw.local.4
+
+    exec.reorder_stack_words
+
+    # --- #
+
+    exec.gen_four_message_words
+
+    popw.local.4
+    movupw.2
+
+    pushw.local.1
+    drop
+    repeat.2
+        swap
+        drop
+    end
+
+    popw.mem # write to mem msg[20, 21, 22, 23] #
+    pushw.local.4
+
+    exec.reorder_stack_words
+
+    # --- #
+
+    exec.gen_four_message_words
+
+    popw.local.4
+    movupw.2
+
+    pushw.local.1
+    drop
+    drop
+    swap
+    drop
+
+    popw.mem # write to mem msg[24, 25, 26, 27] #
+    pushw.local.4
+
+    exec.reorder_stack_words
+
+    # --- #
+
+    exec.gen_four_message_words
+
+    popw.local.4
+    movupw.2
+
+    pushw.local.1
+    drop
+    drop
+    drop
+
+    popw.mem # write to mem msg[28, 29, 30, 31] #
+    pushw.local.4
+
+    exec.reorder_stack_words
+
+    # --- #
+    # --- #
+
+    exec.gen_four_message_words
+
+    popw.local.4
+    movupw.2
+
+    pushw.local.2
+    repeat.3
+        swap
+        drop
+    end
+
+    popw.mem # write to mem msg[32, 33, 34, 35] #
+    pushw.local.4
+
+    exec.reorder_stack_words
+
+    # --- #
+
+    exec.gen_four_message_words
+
+    popw.local.4
+    movupw.2
+
+    pushw.local.2
+    drop
+    repeat.2
+        swap
+        drop
+    end
+
+    popw.mem # write to mem msg[36, 37, 38, 39] #
+    pushw.local.4
```

Comment:
This could be replaced by something like:
```
storew.mem
loadw.local.4
```
This change should save 8 cycles.

Reply:
Noted